Upgrade Kubelet after APIServer

alexandre-allard opened the pull request:
**Component**: salt, scripts

**Context**: 
Upgrade fails because we upgrade Kubelet version before APIServer one, we end up with a kubelet version superior than the APIServer, which is not supported.

**Summary**:
Ensure all APIServer are upgraded to the new version before upgrading anything else.
We probably want to do the opposite for the downgrade (downgrade Kubelet on all nodes first or rather downgrade all APIServer instances last).
We may also want to do another bootstrap role instead of using a pillar entry.

**Acceptance criteria**: 
Upgrade from 2.5.1  to 2.6.0-dev works

---

Closes: #2811

alexandre-allard committed:
scripts,docs: remove destination version option

This option is no longer needed since we
must always use the script shipped with the
destination version ISO, so the version can
be rendered at build time into the script.

alexandre-allard committed:
salt: upgrade all apiserver instances after etcd

We now upgrade all apiserver instances right
after etcd, before upgrading the rest of the
components since apiserver must be upgraded
before CM, scheduler and kubelet.

Refs: #2811

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2811-kubelet-upgraded-before-apiserver` into
`development/2.6`
* [w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

gdemonet commented on scripts/upgrade.sh.in:
```diff
@@ -6,17 +6,17 @@ set -o pipefail
 _usage() {
     echo "upgrade.sh [options]"
     echo "Options:"
-    echo "--destination-version "
-    echo "   <destination-version>:        Destination version to upgrade to"
```
Not really related to this PR, is it? Also, would need to update documentation accordingly.

alexandre-allard replied:
already done, not pushed yet.

gdemonet commented on salt/metalk8s/orchestrate/apiserver.sls:
```diff
@@ -0,0 +1,39 @@
+{%- set dest_version = pillar.metalk8s.cluster_version %}
```
Given you're passing it from the call site, and that it may likely be useful in case of downgrade (I guess), maybe we should consume it:

```suggestion
{%- set dest_version = pillar.orchestrate.dest_version %}
```

alexandre-allard replied:
bad copy/paste. :)

gdemonet commented on salt/metalk8s/roles/bootstrap/local.sls:
```diff
@@ -1,6 +1,8 @@
 include:
   - metalk8s.archives.mounted
+{%- if not pillar.skip_kubelet | default(False) %}
```
Didn't we settle on having another "role" to only upgrade the Salt master? I don't know, maybe simply running `state.sls metalk8s.salt.master` from the upgrade script would be enough?

TeddyAndrieux replied:
(And repo, we want to upgrade repo as well at this point)

No strong opinion here for me either:
1- Do like this with pillar variable (I would just move it to `metalk8s.skip.kubelet`)
2- Create a new role state in bootstrap (e.g.: `metalk8s.roles.bootstrap.component`) that only do salt.master and repo and then include this state here
3- Call directly the right state in the upgrade script (but may be a bit harder to maintain in the futur)

NOTE: Also to me you also want to skip, at least, kubectl during upgrade of bootstrap (kubectl should be upgraded after apiserver same as kubelet)

gdemonet replied:
IMO, skipping (option 1) and listing which states to use (option 3) are equally hard to maintain, as they force some coupling between the script and the states (to apply, or not to apply).
I guess this leaves us with option 2 :slightly_smiling_face:

alexandre-allard replied:
Repositories are already upgraded as well as Salt master kubeconfig and certs because of the `metalk8s.orchestrate.bootstrap.pre-upgrade` right before, so we only need to run `metalk8s.salt.master.installed` to make it works.
Maybe it's better if the repositories are explicitly deployed in some new role as you suggested, instead of relying on some dependencies, I will change it.

TeddyAndrieux replied:
> Repositories are already upgraded as well as Salt master kubeconfig and certs because of the metalk8s.orchestrate.bootstrap.pre-upgrade right before

Are you sure ? We only update the yum repositories configs on the host but not the repo container itself

alexandre-allard-scality replied:
My bad, you're right, it's even earlier in the process.
It's due to the iso-manager script.

alexandre-allard replied:
My bad, you're right, just the iso manager popuplate the registry and salt master with stuff from the new ISO, it does not upgrade yet the images, still the new content is accessible.

gdemonet reviewed (commented):
Some early comments :)

bert-e commented:
# History mismatch



Merge commit #741479ce5d672efc2b00a8b57387e14a9d023df0 on the integration branch
`w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver` is merging a branch which is neither the current
branch `bugfix/2811-kubelet-upgraded-before-apiserver` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `bugfix/2811-kubelet-upgraded-before-apiserver` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard committed:
scripts: only upgrade salt master and registry on bootstrap

We now only upgrade salt master, containers
registry and repositories on the bootstrap,
They are the only components we need to upgrade
prior to the upgrade orchestrate to make new Salt
modules & formulas, packages and container images
available to the cluster.

Refs: #2811

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2811-kubelet-upgraded-before-apiserver` into
`development/2.6`
* [w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

TeddyAndrieux reviewed (dismissed):
Great :ok_hand: 
Not tested, but changes LGTM

TeddyAndrieux commented on salt/metalk8s/salt/master/files/salt-master-manifest.yaml.j2:
```diff
@@ -143,3 +145,7 @@ spec:
       hostPath:
         path: '/etc/kubernetes/pki'
         type: Directory
+    - name: salt-master-log
+      hostPath:
+        path: '/var/log/salt/master'
+        type: FileOrCreate
```
To be consistent with other I would add the full directory as it's the directory for salt it does not hurt IMO

Also not sure it's really relevant but to be sure maybe do the same for salt api

gdemonet commented on salt/metalk8s/salt/master/files/salt-master-manifest.yaml.j2:
```diff
@@ -78,6 +78,8 @@ spec:
         - name: kubernetes-pki
           mountPath: '/etc/kubernetes/pki'
           readOnly: true
+        - name: salt-master-log
+          mountPath: '/var/log/salt/master'
```
Isn't salt-master logging to `stdout`/`stderr` when running in foreground mode?

> It makes easier to read and debug salt-master

Isn't `kubectl logs ...` sufficient?

alexandre-allard replied:
I've removed the commit.
But I'm not sure you can have the full history from kubectl or crictl.

alexandre-allard replied:
One could argue we already have Loki for that, but Loki could be down.
Moreover, from my PoV, it's much simpler to have the log directly accessible on your host (to grep, etc.), rather than having to go on a UI.
But I agree, that's not 100% necessary.

gdemonet replied:
You can `grep` on the output of `kubectl logs` :wink:
If we start bypassing the container logging approach to get them on the host, we'll have to maintain our own rotation, monitor the space it takes, etc.
And for the log history, we have Loki indeed :) (which should be usable without a UI FWIW)

bert-e commented:
# History mismatch



Merge commit #f727eee0eefb35cddd29491856f09543e222175a on the integration branch
`w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver` is merging a branch which is neither the current
branch `bugfix/2811-kubelet-upgraded-before-apiserver` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `bugfix/2811-kubelet-upgraded-before-apiserver` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2811-kubelet-upgraded-before-apiserver` into
`development/2.6`
* [w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

alexandre-allard-scality commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/7ab54542543377ab68b2d4ea15c6b5ac4d1bcbb6) did not succeed in branch w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #166e479c939b2138ecc28017570208663d83b556 on the integration branch
`w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver` is merging a branch which is neither the current
branch `bugfix/2811-kubelet-upgraded-before-apiserver` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `bugfix/2811-kubelet-upgraded-before-apiserver` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2811-kubelet-upgraded-before-apiserver` into
`development/2.6`
* [w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal










*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #a719563b0138646e98c4c80ca6e0c067f56c5587 on the integration branch
`w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver` is merging a branch which is neither the current
branch `bugfix/2811-kubelet-upgraded-before-apiserver` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `bugfix/2811-kubelet-upgraded-before-apiserver` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2811-kubelet-upgraded-before-apiserver` into
`development/2.6`
* [w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/bugfix/2811-kubelet-upgraded-before-apiserver) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

alexandre-allard commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
